From 3d9fcd683751c31a2abfdeebb704c37caabbd5cc Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Thu, 16 Apr 2020 12:04:09 +0800
Subject: [PATCH 31/68] MMFMWK-8759 waylandsink: add listener for wl_point and
 wl_touch

Add listener for wl_point and wl_touch, these handler
can make surface respond to mouse and touch event.

Upstream-Status: Pending
https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/1195

(cherry picked from commit 7714b8f0f148acf94f34317e1cadebeae845205c)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/wayland/gstwldisplay.c |  55 ++++++++++++
 gst-libs/gst/wayland/gstwldisplay.h |   9 ++
 gst-libs/gst/wayland/gstwlwindow.c  | 129 +++++++++++++++++++++++++++-
 3 files changed, 192 insertions(+), 1 deletion(-)

diff --git a/gst-libs/gst/wayland/gstwldisplay.c b/gst-libs/gst/wayland/gstwldisplay.c
index 1938d86..e3e2c21 100644
--- a/gst-libs/gst/wayland/gstwldisplay.c
+++ b/gst-libs/gst/wayland/gstwldisplay.c
@@ -49,6 +49,9 @@ typedef struct _GstWlDisplayPrivate
   struct wl_compositor *compositor;
   struct wl_subcompositor *subcompositor;
   struct xdg_wm_base *xdg_wm_base;
+  struct wl_seat *seat;
+  struct wl_pointer *pointer;
+  struct wl_touch *touch;
   struct zwp_fullscreen_shell_v1 *fullscreen_shell;
   struct wp_single_pixel_buffer_manager_v1 *single_pixel_buffer;
   struct wl_shm *shm;
@@ -309,6 +312,32 @@ gst_wl_display_check_format_for_dmabuf (GstWlDisplay * self,
   return FALSE;
 }
 
+static void
+seat_handle_capabilities (void *data, struct wl_seat *seat,
+    enum wl_seat_capability caps)
+{
+  GstWlDisplay *self = data;
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  if ((caps & WL_SEAT_CAPABILITY_POINTER) && !priv->pointer) {
+    priv->pointer = wl_seat_get_pointer (seat);
+  } else if (!(caps & WL_SEAT_CAPABILITY_POINTER) && priv->pointer) {
+    wl_pointer_destroy (priv->pointer);
+    priv->pointer = NULL;
+  }
+
+  if ((caps & WL_SEAT_CAPABILITY_TOUCH) && !priv->touch) {
+    priv->touch = wl_seat_get_touch (seat);
+  } else if (!(caps & WL_SEAT_CAPABILITY_TOUCH) && priv->touch) {
+    wl_touch_destroy (priv->touch);
+    priv->touch = NULL;
+  }
+}
+
+static const struct wl_seat_listener seat_listener = {
+  seat_handle_capabilities,
+};
+
 static void
 handle_xdg_wm_base_ping (void *user_data, struct xdg_wm_base *xdg_wm_base,
     uint32_t serial)
@@ -381,6 +410,9 @@ registry_handle_global (void *data, struct wl_registry *registry,
     priv->xdg_wm_base =
         wl_registry_bind (registry, id, &xdg_wm_base_interface, 1);
     xdg_wm_base_add_listener (priv->xdg_wm_base, &xdg_wm_base_listener, self);
+  } else if (strcmp (interface, "wl_seat") == 0) {
+    priv->seat = wl_registry_bind (registry, id, &wl_seat_interface, 1);
+    wl_seat_add_listener (priv->seat, &seat_listener, self);
   } else if (g_strcmp0 (interface, "zwp_fullscreen_shell_v1") == 0) {
     priv->fullscreen_shell = wl_registry_bind (registry, id,
         &zwp_fullscreen_shell_v1_interface, 1);
@@ -691,6 +723,13 @@ gst_wl_display_get_xdg_wm_base (GstWlDisplay * self)
   return priv->xdg_wm_base;
 }
 
+struct wl_seat *gst_wl_display_get_seat (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->seat;
+}
+
 struct zwp_fullscreen_shell_v1 *
 gst_wl_display_get_fullscreen_shell_v1 (GstWlDisplay * self)
 {
@@ -715,6 +754,22 @@ gst_wl_display_get_alpha_compositing (GstWlDisplay * self)
   return priv->alpha_compositing;
 }
 
+struct wl_pointer *
+gst_wl_display_get_pointer (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->pointer;
+}
+
+struct wl_touch *
+gst_wl_display_get_touch (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->touch;
+}
+
 gint
 gst_wl_display_get_width (GstWlDisplay * self)
 {
diff --git a/gst-libs/gst/wayland/gstwldisplay.h b/gst-libs/gst/wayland/gstwldisplay.h
index 2d34567..9eae9f8 100644
--- a/gst-libs/gst/wayland/gstwldisplay.h
+++ b/gst-libs/gst/wayland/gstwldisplay.h
@@ -85,6 +85,9 @@ struct wl_subcompositor *gst_wl_display_get_subcompositor (GstWlDisplay * self);
 GST_WL_API
 struct xdg_wm_base *gst_wl_display_get_xdg_wm_base (GstWlDisplay * self);
 
+GST_WL_API
+struct wl_seat *gst_wl_display_get_seat (GstWlDisplay * self);
+
 GST_WL_API
 struct zwp_fullscreen_shell_v1 *gst_wl_display_get_fullscreen_shell_v1 (GstWlDisplay * self);
 
@@ -94,6 +97,12 @@ struct wp_viewporter *gst_wl_display_get_viewporter (GstWlDisplay * self);
 GST_WL_API
 struct zwp_alpha_compositing_v1 *gst_wl_display_get_alpha_compositing (GstWlDisplay * self);
 
+GST_WL_API
+struct wl_pointer *gst_wl_display_get_pointer (GstWlDisplay * self);
+
+GST_WL_API
+struct wl_touch *gst_wl_display_get_touch (GstWlDisplay * self);
+
 GST_WL_API
 gint gst_wl_display_get_width (GstWlDisplay * self);
 
diff --git a/gst-libs/gst/wayland/gstwlwindow.c b/gst-libs/gst/wayland/gstwlwindow.c
index 8ee8dcf..6b9f8e1 100644
--- a/gst-libs/gst/wayland/gstwlwindow.c
+++ b/gst-libs/gst/wayland/gstwlwindow.c
@@ -24,6 +24,8 @@
 #include <config.h>
 #endif
 
+#include <linux/input.h>
+
 #include "gstwlwindow.h"
 #include "gstwlutils.h"
 #include "gstimxcommon.h"
@@ -92,6 +94,9 @@ typedef struct _GstWlWindowPrivate
 
   /* video buffer scale */
   guint scale;
+
+  /* mouse location when click */
+  gint pointer_x, pointer_y;
 } GstWlWindowPrivate;
 
 G_DEFINE_TYPE_WITH_CODE (GstWlWindow, gst_wl_window, G_TYPE_OBJECT,
@@ -100,6 +105,9 @@ G_DEFINE_TYPE_WITH_CODE (GstWlWindow, gst_wl_window, G_TYPE_OBJECT,
         "wlwindow", 0, "wlwindow library");
     );
 
+/* resize trigger margin in pixel */
+#define RESIZE_MARGIN 20
+
 enum
 {
   CLOSED,
@@ -116,6 +124,110 @@ static void gst_wl_window_update_borders (GstWlWindow * self);
 static void gst_wl_window_commit_buffer (GstWlWindow * self,
     GstWlBuffer * buffer);
 
+static void
+pointer_handle_enter (void *data, struct wl_pointer *pointer,
+    uint32_t serial, struct wl_surface *surface, wl_fixed_t sx, wl_fixed_t sy)
+{
+  GstWlWindow *self = data;
+  GstWlWindowPrivate *priv = gst_wl_window_get_instance_private (self);
+
+  priv->pointer_x = wl_fixed_to_int (sx);
+  priv->pointer_y = wl_fixed_to_int (sy);
+}
+
+static void
+pointer_handle_leave (void *data, struct wl_pointer *pointer,
+    uint32_t serial, struct wl_surface *surface)
+{
+}
+
+static void
+pointer_handle_motion (void *data, struct wl_pointer *pointer,
+    uint32_t time, wl_fixed_t sx, wl_fixed_t sy)
+{
+}
+
+static void
+pointer_handle_button (void *data, struct wl_pointer *wl_pointer,
+    uint32_t serial, uint32_t time, uint32_t button, uint32_t state)
+{
+  GstWlWindow *self = data;
+  GstWlWindowPrivate *priv = gst_wl_window_get_instance_private (self);
+
+  if (!priv->xdg_toplevel)
+    return;
+
+  if (button == BTN_LEFT && state == WL_POINTER_BUTTON_STATE_PRESSED) {
+    struct wl_seat *seat = gst_wl_display_get_seat (priv->display);
+    if (priv->render_rectangle.w - priv->pointer_x <= RESIZE_MARGIN
+        && priv->render_rectangle.h - priv->pointer_y <= RESIZE_MARGIN)
+      xdg_toplevel_resize (priv->xdg_toplevel, seat, serial,
+          XDG_TOPLEVEL_RESIZE_EDGE_BOTTOM_RIGHT);
+    else
+      xdg_toplevel_move (priv->xdg_toplevel, seat, serial);
+  }
+}
+
+static void
+pointer_handle_axis (void *data, struct wl_pointer *wl_pointer,
+    uint32_t time, uint32_t axis, wl_fixed_t value)
+{
+}
+
+static const struct wl_pointer_listener pointer_listener = {
+  pointer_handle_enter,
+  pointer_handle_leave,
+  pointer_handle_motion,
+  pointer_handle_button,
+  pointer_handle_axis,
+};
+
+static void
+touch_handle_down (void *data, struct wl_touch *wl_touch,
+    uint32_t serial, uint32_t time, struct wl_surface *surface,
+    int32_t id, wl_fixed_t x_w, wl_fixed_t y_w)
+{
+  GstWlWindow *self = data;
+  GstWlWindowPrivate *priv = gst_wl_window_get_instance_private (self);
+  struct wl_seat *seat;
+
+  if (!priv->xdg_toplevel)
+    return;
+
+  seat = gst_wl_display_get_seat (priv->display);
+  xdg_toplevel_move (priv->xdg_toplevel, seat, serial);
+}
+
+static void
+touch_handle_up (void *data, struct wl_touch *wl_touch,
+    uint32_t serial, uint32_t time, int32_t id)
+{
+}
+
+static void
+touch_handle_motion (void *data, struct wl_touch *wl_touch,
+    uint32_t time, int32_t id, wl_fixed_t x_w, wl_fixed_t y_w)
+{
+}
+
+static void
+touch_handle_frame (void *data, struct wl_touch *wl_touch)
+{
+}
+
+static void
+touch_handle_cancel (void *data, struct wl_touch *wl_touch)
+{
+}
+
+static const struct wl_touch_listener touch_listener = {
+  touch_handle_down,
+  touch_handle_up,
+  touch_handle_motion,
+  touch_handle_frame,
+  touch_handle_cancel,
+};
+
 static void
 handle_xdg_toplevel_close (void *data, struct xdg_toplevel *xdg_toplevel)
 {
@@ -130,6 +242,7 @@ handle_xdg_toplevel_configure (void *data, struct xdg_toplevel *xdg_toplevel,
     int32_t width, int32_t height, struct wl_array *states)
 {
   GstWlWindow *self = data;
+  GstWlWindowPrivate *priv = gst_wl_window_get_instance_private (self);
   const uint32_t *state;
 
   GST_DEBUG ("XDG toplevel got a \"configure\" event, [ %d, %d ].",
@@ -145,10 +258,12 @@ handle_xdg_toplevel_configure (void *data, struct xdg_toplevel *xdg_toplevel,
     }
   }
 
-  if (width <= 0 || height <= 0)
+  if (width <= 2 * RESIZE_MARGIN || height <= 2 * RESIZE_MARGIN)
     return;
 
+  g_mutex_lock (priv->render_lock);
   gst_wl_window_set_render_rectangle (self, 0, 0, width, height);
+  g_mutex_unlock (priv->render_lock);
 }
 
 static const struct xdg_toplevel_listener xdg_toplevel_listener = {
@@ -349,6 +464,8 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
   /* Check which protocol we will use (in order of preference) */
   if (xdg_wm_base) {
     gint64 timeout;
+    struct wl_pointer *pointer;
+    struct wl_touch *touch;
 
     /* First create the XDG surface */
     priv->xdg_surface = xdg_wm_base_get_xdg_surface (xdg_wm_base,
@@ -373,6 +490,16 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
       xdg_toplevel_set_app_id (priv->xdg_toplevel, "org.gstreamer.wayland");
     }
 
+    pointer = gst_wl_display_get_pointer (display);
+    touch = gst_wl_display_get_touch (display);
+    if (pointer)
+      wl_pointer_add_listener (pointer, &pointer_listener, self);
+
+    if (touch) {
+      wl_touch_set_user_data (touch, self);
+      wl_touch_add_listener (touch, &touch_listener, self);
+    }
+
     gst_wl_window_ensure_fullscreen (self, fullscreen);
 
     /* Finally, commit the xdg_surface state as toplevel */
-- 
2.43.0

